File: react/aaudiocore/src/strategies/shuffle-strategy.ts
import type { IPlaybackStrategy } from "../contracts/playback-strategy";
import type { PlaybackTrack } from "../models/track";

export class ShuffleStrategy implements IPlaybackStrategy {
  readonly name = "shuffle";

  selectNext(currentIndex: number, queue: PlaybackTrack[]): number {
    if (queue.length === 0) return -1;
    if (queue.length === 1) return 0;

    let nextIndex = currentIndex;
    while (nextIndex === currentIndex) {
      nextIndex = Math.floor(Math.random() * queue.length);
    }

    return nextIndex;
  }
}

File: winforms/aaudiocore/Strategies/ShuffleStrategy.cs
using AAudioCore.Contracts;
using AAudioCore.Models;

namespace AAudioCore.Strategies;

public sealed class ShuffleStrategy : IPlaybackStrategy
{
    private readonly Random random = new();

    public string Name => "shuffle";

    public int SelectNext(int currentIndex, IReadOnlyList<PlaybackTrack> queue)
    {
        if (queue.Count == 0) return -1;
        if (queue.Count == 1) return 0;

        var nextIndex = currentIndex;
        while (nextIndex == currentIndex)
        {
            nextIndex = random.Next(queue.Count);
        }

        return nextIndex;
    }
}
